auth_core/pkce.py

```diff
@@ -0,0 +1,50 @@
+import base64
+import hashlib
+import httpx
+import logging
+import secrets
+
+from urllib.parse import urljoin
+
+from .token_data import TokenData
+
+
+logger = logging.getLogger("gel_auth_core")
+
+
+class PKCE:
+    def __init__(self, verifier: str, *, base_url: str):
+        self.base_url = base_url
+        self.verifier = verifier
+        self.challenge = (
+            base64.urlsafe_b64encode(hashlib.sha256(verifier.encode()).digest())
+            .rstrip(b"=")
+            .decode()
+        )
+
+    async def exchange_code_for_token(self, code: str) -> TokenData:
+        async with httpx.AsyncClient() as http_client:
+            url = urljoin(self.base_url, "token")
+            logger.info(f"Exchanging code for token: {url}")
```

Comment:
```suggestion
            logger.info("Exchanging code for token: %s", url)
```

For potentially less memory copies fwiw.